[WFCORE-4227] Add the ability for the CLI SSL security commands to be a…

MMarus opened the pull request:
…ble to obtain a server certificate from Let's Encrypt

WFCORE issue: https://issues.jboss.org/browse/WFCORE-4227
EAP issue: https://issues.jboss.org/browse/EAP7-1100
Analysis document: wildfly/wildfly-proposals#156
Community documentation: wildfly/wildfly#11921

Please do not merge before https://issues.jboss.org/browse/EAP7-1150 is fixed.

Add the ability for the CLI SSL security commands to be able to obtain a server certificate from Let's Encrypt.

MMarus commented on elytron/src/main/java/org/wildfly/extension/elytron/CertificateAuthorityAccountDefinition.java:
```diff
@@ -78,10 +78,11 @@
     enum CertificateAuthority {
-
-        LETS_ENCRYPT("LetsEncrypt", "https://acme-v02.api.letsencrypt.org/" + DIRECTORY, "https://acme-staging-v02.api.letsencrypt.org/" + DIRECTORY);
+        //TODO: replace with the configuration
```
I will remove changes in this file once the https://issues.jboss.org/browse/EAP7-1150 is fixed.

MMarus commented:
The test CertificateAuthoritiesTestCase.testDeactivateAccount fails because the changes in "CertificateAuthorityAccountDefinition.java" which are going to be removed when the   https://issues.jboss.org/browse/EAP7-1150 is fixed.

darranl commented:
Can we please ensure the commit messages and title of the pull request only reference the WFCORE Jira issues?  

These both end up in the commit history in the repo and are the only publicly accessible Jira issues.

jfdenise commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/InteractiveSecurityBuilder.java:
```diff
@@ -376,6 +593,12 @@ protected KeyStore buildKeyStore(CommandContext ctx, boolean buildRequest) throw
         if (keyStoreName != null) {
+            if(useLetsEncrypt && caAccount == null) {
+                ModelNode req = ElytronUtil.removeCertificateAuthorityAccount(ctx, certAuthorityAccountName);
+                SecurityCommand.execute(ctx, req, SecurityCommand.DEFAULT_FAILURE_CONSUMER);
```
Be sure to call execute with noReload argument to be false, otherwise the server stays in reload state.

jfdenise commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/InteractiveSecurityBuilder.java:
```diff
@@ -376,6 +593,12 @@ protected KeyStore buildKeyStore(CommandContext ctx, boolean buildRequest) throw
+                ModelNode req = ElytronUtil.removeCertificateAuthorityAccount(ctx, certAuthorityAccountName);
+                SecurityCommand.execute(ctx, req, SecurityCommand.DEFAULT_FAILURE_CONSUMER);
+                req = ElytronUtil.removeKeyStore(ctx, accountKeyStoreName);
+                SecurityCommand.execute(ctx, req, SecurityCommand.DEFAULT_FAILURE_CONSUMER);
```
Be sure to call execute with noReload argument to be false, otherwise the server stays in reload state.

MMarus commented:
I have also created the tests for the "enable-ssl-http-server" in the wildfly [1] and they passed in the CI [2]. I will create new PR in the wildfly with the test for the when wildfly dependency on core is upgraded to core that contains this PR.

[1] https://github.com/MMarus/wildfly/tree/WFCORE-4227
[2] https://ci.wildfly.org/viewLog.html?order=DURATION_DESC&recordsPerPage=20&filterText=SecurityCommands&status=&currentGroup=test&scope=%23teamcity%23org.jboss.as.test.integration.management.cli%23teamcity%23SecurityCommandsTestCase&pager.currentPage=1&buildTypeId=WF_WildFlyCoreIntegrationExperiments&buildId=134166&tab=testsInfo

MMarus commented:
Retest this please.

MMarus commented:
Retest this please!

MMarus commented:
I've added tests mentioned in test plan and updated the implementation to use changes in the feature branch for EAP7-1150.

jamezp commented:
This is more for future reference, but we should probably open PR's like this as drafts. Maybe at the time this was filed drafts weren't even an option :)

I'm placing a hold on this since it requires a SNAPSHOT dependency.

darranl commented:
We haven't reached the point that Diana's PR is reviewed, at this stage probably better to not include those changes in a follow on PR - if Diana makes changes to her PR as a result of review it may change here.

darranl commented on pom.xml:
```diff
@@ -216,7 +216,7 @@
         <version.org.wildfly.openssl.wildfly-openssl-windows-i386>${version.org.wildfly.openssl.natives}</version.org.wildfly.openssl.wildfly-openssl-windows-i386>
         <version.org.wildfly.openssl.wildfly-openssl-windows-x86_64>${version.org.wildfly.openssl.natives}</version.org.wildfly.openssl.wildfly-openssl-windows-x86_64>
-        <version.org.wildfly.security.elytron>1.9.0.CR4</version.org.wildfly.security.elytron>
+        <version.org.wildfly.security.elytron>1.9.0.CR5-SNAPSHOT</version.org.wildfly.security.elytron>
```
Overall when submitting a PR please never include the dependency on the SNAPSHOT version of the component, it makes it impossible to coordinate the merging of this PR as first that needs to be undone.

Overall please just submit the changes relevant to your enhancement.

darranl commented:
Also when adding binary keystores to the testsuite can you please ensure they are accompanied with a README.txt describing what is in them and any passwords required to access them.

It should also be considered if these can be dynamically generated at build time to avoid issues in relation to the expiration of certificates at some point in the future.

darranl commented:
@jamezp +1 to not including SNAPSHOT dependencies in the pom, however as a feature submitting as a draft PR would not be suitable.  When PRs are submitted as a draft it disables the ability to review the PR, however as soon as feature PRs are submitted they should be ready for code review.

As a five month old PR the majority of code review should be complete by now, the only delay on this one being the feature it depends on being developed in parallel.

MMarus commented:
@darranl @jamezp thank you very much for review. I've removed the SNAPSHOT dependency and the Diana's changes, however the CI builds are gonna fail until the https://github.com/wildfly/wildfly-core/pull/3768 is not merged.

I have also added the README.txt for the keystores.

darranl commented:
Thanks @MMarus the CI failures are unfortunately something to live with in this process at the moment, we have some ideas hopefully for the future to "pipeline" enhancements and test all pieces together.  

Once Diana's PR is in I will kick off the CI jobs again which should then mean this PR has everything it needs.

vmarekrh commented on testsuite/shared/src/main/java/org/jboss/as/test/integration/management/cli/AcmeMockServerBuilder.java:
```diff
@@ -0,0 +1,560 @@
+
+        final String CHECK_ORDER_RESPONSE_BODY = "{" + System.lineSeparator() +
+                "  \"status\": \"valid\"," + System.lineSeparator() +
+                "  \"expires\": \"2018-05-04T18:27:35Z\"," + System.lineSeparator() +
```
It is valid to use already expired response?
Does it check the time of expiring?

If yes, it doesn't good to use current time + one day to expire time?

fjuma replied:
The "expires" field here is used by the ACME server (i.e., certificate authority) only. It is a timestamp after which the ACME server will consider the certificate order to be invalid. The mock ACME server that is being used in the tests makes use of pre-recorded messages which is why it doesn't fail even though the date is in the past.

vmarekrh commented on testsuite/shared/src/main/java/org/jboss/as/test/integration/management/cli/AcmeMockServerBuilder.java:
```diff
@@ -0,0 +1,560 @@
+
+        final String FINALIZE_RESPONSE_BODY = "{" + System.lineSeparator() +
+                "  \"status\": \"valid\"," + System.lineSeparator() +
+                "  \"expires\": \"2018-05-04T18:27:35Z\"," + System.lineSeparator() +
```
It is valid to use already expired response?
Does it check the time of expiring?

If yes, it doesn't good to use current time + one day to expire time?

MMarus replied:
No it does not check the expiration date since it is only a mock server which is not able to process the answers as normal ACME server would. It just returns pregenerated strings. If we would change it dynamically then the expected requests would need to be changed dynamically as well which can be done by full ACME server not by a simple mock like this.

vmarekrh replied:
I didn't mean the moc server will validate the message what he sends, but I mean if the wildfly-core server validates the time of expiring of the received message from ACME server.
If the wildfly-core server validates the whole received message from ACME server.

If the wildfly-core didn't validate the answer of ACME server, isn't it the hole?

MMarus replied:
That is a good point, I am not sure whether the Elytron does these checks when communicating with ACME server. Maybe @fjuma can give you a better answer?

fjuma replied:
The "expires" field here is used by the ACME server (i.e., certificate authority) only. It is a timestamp after which the ACME server will consider the proof provided by the owner of the certificate to be invalid. The mock ACME server that is being used in the tests makes use of pre-recorded messages which is why it doesn't fail even though the date is in the past.

vmarekrh commented on testsuite/standalone/src/test/java/org/jboss/as/test/integration/management/cli/SecurityCommandsTestCase.java:
```diff
@@ -77,11 +97,271 @@
+        testEnableLetsEncryptSSLInteractiveConfirm(null, true);
+    }
+
+    private void testEnableLetsEncryptSSLInteractiveConfirm(String mgmtInterface, boolean useCaAccount) throws Exception {
```
Why have this method parameter mgmtInterface? Both tests what it calls with null this parameter.

MMarus replied:
Removed mgmtInterface.

MMarus commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/ElytronUtil.java:
```diff
@@ -1223,4 +1224,73 @@ public static boolean constantRoleMapperExists(CommandContext ctx, String name)
         return Util.isSuccess(ctx.getModelControllerClient().execute(builder.buildRequest()));
     }
+
+    public static ModelNode addCertificateAuthority(CommandContext ctx, CertificateAuthority certificateAuthority) throws Exception {
```
removed the ctx

MMarus commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/ElytronUtil.java:
```diff
@@ -1223,4 +1224,73 @@ public static boolean constantRoleMapperExists(CommandContext ctx, String name)
+        return builder.buildRequest();
+    }
+
+    public static ModelNode addCertificateAuthorityAccount(CommandContext ctx, String name, String password, String alias, String keyStoreName, List<String> contactUrls, CertificateAuthority customCertificateAuthority) throws Exception {
```
removed the ctx

MMarus commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/ElytronUtil.java:
```diff
@@ -1223,4 +1224,73 @@ public static boolean constantRoleMapperExists(CommandContext ctx, String name)
+        return modelNodes;
+    }
+
+    public static ModelNode removeCertificateAuthorityAccount(CommandContext ctx, String name) throws Exception {
```
removed the ctx

MMarus commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/ElytronUtil.java:
```diff
@@ -1223,4 +1224,73 @@ public static boolean constantRoleMapperExists(CommandContext ctx, String name)
+        return builder.buildRequest();
+    }
+
+    public static ModelNode obtainCertificateRequest(CommandContext ctx,
```
removed the ctx

MMarus commented on testsuite/standalone/src/test/java/org/jboss/as/test/integration/management/cli/SecurityCommandsTestCase.java:
```diff
@@ -77,11 +97,271 @@
+            Assert.assertTrue(cli.pushLineAndWaitForResults(CA_ACCOUNT_ALIAS, "Certificate authority URL (default "+ certificateAuthority.getUrl() +"):"));
+            Assert.assertTrue(cli.pushLineAndWaitForResults(LOCALHOST_ACME_URL, "Do you agree to Let's Encrypt terms of service? y/n:"));
+            //Check that the interactive mode exits after rejecting the TOS
+            Assert.assertTrue(cli.pushLineAndWaitForResults("n", null));
```
Added error message check.

MMarus commented on testsuite/standalone/src/test/java/org/jboss/as/test/integration/management/cli/SecurityCommandsTestCase.java:
```diff
@@ -77,11 +97,271 @@
+            Assert.assertTrue(cli.pushLineAndWaitForResults(certificateAlias, "Enable SSL Mutual Authentication"));
+            Assert.assertTrue(cli.pushLineAndWaitForResults("n", "Do you confirm"));
+            Assert.assertTrue(cli.pushLineAndWaitForResults("y", null));
+
```
Added error message check.

vmarekrh commented on testsuite/standalone/src/test/java/org/jboss/as/test/integration/management/cli/SecurityCommandsTestCase.java:
```diff
@@ -77,11 +97,266 @@
+            Assert.assertTrue(cli.pushLineAndWaitForResults(certificateAlias, "Enable SSL Mutual Authentication"));
+            Assert.assertTrue(cli.pushLineAndWaitForResults("n", "Do you confirm"));
+            Assert.assertTrue(cli.pushLineAndWaitForResults("y", null));
+            Assert.assertTrue(cli.getOutput().contains("FakeCaAccount not found"));
```
This is not the problem for this pull request.
But while I review the test with fake Ca Account, I discover in the server log word service twice.
Does anyone have an idea why is word service in the exception message written twice?


````
6:39:03,285 ERROR [org.jboss.as.controller.management-operation] (management-handler-thread - 1) WFLYCTL0013: Operation ("obtain-certificate") failed - address: ([
    ("subsystem" => "elytron"),
    ("key-store" => "key-store-f7d7a87a-9df1-4591-8a1d-5cd2d775a9f0")
]): org.jboss.msc.service.ServiceNotFoundException: Service service org.wildfly.security.certificate-authority-account.FakeCaAccount not found
	at org.jboss.msc.service.ServiceContainerImpl.getRequiredService(ServiceContainerImpl.java:652)
	at org.jboss.as.controller.OperationContextImpl$OperationContextServiceRegistry.getRequiredService(OperationContextImpl.java:2282)
	at org.wildfly.extension.elytron.ElytronExtension.getRequiredService(ElytronExtension.java:138)
	at org.wildfly.extension.elytron.AdvancedModifiableKeyStoreDecorator.getAcmeAccount(AdvancedModifiableKeyStoreDecorator.java:1015)
	at org.wildfly.extension.elytron.AdvancedModifiableKeyStoreDecorator.access$100(AdvancedModifiableKeyStoreDecorator.java:104)
	at org.wildfly.extension.elytron.AdvancedModifiableKeyStoreDecorator$ObtainCertificateHandler.executeRuntimeStep(AdvancedModifiableKeyStoreDecorator.java:786)
	at org.jboss.as.controller.AbstractRuntimeOnlyHandler$1.execute(AbstractRuntimeOnlyHandler.java:59)
	at org.jboss.as.controller.AbstractOperationContext.executeStep(AbstractOperationContext.java:999)
	at org.jboss.as.controller.AbstractOperationContext.processStages(AbstractOperationContext.java:743)
	at org.jboss.as.controller.AbstractOperationContext.executeOperation(AbstractOperationContext.java:467)
	at org.jboss.as.controller.OperationContextImpl.executeOperation(OperationContextImpl.java:1412)
	at org.jboss.as.controller.ModelControllerImpl.internalExecute(ModelControllerImpl.java:423)
	at org.jboss.as.controller.ModelControllerImpl.lambda$execute$1(ModelControllerImpl.java:243)
	at org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:289)
	at org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:255)
	at org.jboss.as.controller.ModelControllerImpl.execute(ModelControllerImpl.java:243)
	at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler.doExecute(ModelControllerClientOperationHandler.java:240)
	at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler.access$400(ModelControllerClientOperationHandler.java:138)
	at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1$1.run(ModelControllerClientOperationHandler.java:162)
	at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1$1.run(ModelControllerClientOperationHandler.java:158)
	at org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:313)
	at org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:270)
	at org.jboss.as.controller.AccessAuditContext.doAs(AccessAuditContext.java:254)
	at org.jboss.as.controller.AccessAuditContext.doAs(AccessAuditContext.java:225)
	at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1.execute(ModelControllerClientOperationHandler.java:158)
	at org.jboss.as.protocol.mgmt.ManagementRequestContextImpl$1.doExecute(ManagementRequestContextImpl.java:70)
	at org.jboss.as.protocol.mgmt.ManagementRequestContextImpl$AsyncTaskRunner.run(ManagementRequestContextImpl.java:160)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:485)

````

vmarekrh replied:
I created Jira issue for this bug [WFCORE-4515](https://issues.jboss.org/browse/WFCORE-4515)

fjuma commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/InteractiveSecurityBuilder.java:
```diff
@@ -130,14 +151,125 @@ public void buildRequest(CommandContext ctx, boolean buildRequest) throws Except
+
+            //then contactUrls / email
+            while (contactUrls == null || contactUrls.size() < 1) {
+                String emails = commandInvocation.inputLine(new Prompt("Contact email(s) [admin@example.com,info@example.com]: "));
```
Since contact URL(s) are optional, we should allow it to not be specified. It's ok to show example contact URLs so a user knows how to format them but we shouldn't use these as default values.

fjuma commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/InteractiveSecurityBuilder.java:
```diff
@@ -168,23 +300,35 @@ public void buildRequest(CommandContext ctx, boolean buildRequest) throws Except
+            //then prompt for domainName
+            while (domainNames == null || domainNames.size() < 1) {
+                String domains = commandInvocation.inputLine(new Prompt("Your domain name(s) (must be accessible by the Let's Encrypt server at 80 & 443 ports) [example.com,second.example.com]: "));
+                domainNames = parseItemsFromString(domains, "example.com,second.example.com");
```
Showing these examples is ok so user know what format is expected but we shouldn't use these as default values.

fjuma commented on testsuite/shared/src/main/resources/org/jboss/as/test/integration/management/cli/README.txt:
```diff
@@ -0,0 +1,140 @@
+
+
+
+======= test.keystore =======
```
I might have missed it but are the entries in test.keystore actually used by any of the tests?

jfdenise commented on cli/src/main/java/org/jboss/as/cli/impl/aesh/cmd/security/model/InteractiveSecurityBuilder.java:
```diff
@@ -286,9 +451,15 @@ public void buildRequest(CommandContext ctx, boolean buildRequest) throws Except
-                        + " file will be generated in server configuration directory.");
+                if (useLetsEncrypt) {
+                    ctx.printLine("Certificate will be obtained from Let's Encrypt server and will be valid for 90 days." + "\n"
+                            + "Server keystore file a will be generated in server configuration directory.");
```
a will be ==> will be

MMarus commented:
@fjuma I am sorry for the quiet period. I was on PTO. I'll update the PR today :).

MMarus commented:
@jfdenise @fjuma Thank you very much for the reviews. I have addressed your comments in the latest commit. If you prefer to squash the commits I'll do that asap :).
